File: forwindows/src/AnyClipCore/Watchdogs.cs
namespace AnyClip.Core;

/// Thrown by watchdogs to unwind the daemon task set; the in-process
/// supervisor restarts with backoff (Python: RuntimeError → supervisor).
public sealed class DaemonRestartException(string message) : Exception(message);

/// mDNS service control implemented by the platform layer (Windows
/// MdnsBeacon over dnsapi; fakes in tests).
public interface IMdnsService
{
    string? AdvertisedIp { get; }
    Task StartAsync(string instanceName, int port,
        IReadOnlyList<(string Key, string Value)> txt);
    void Refresh();
    void Stop();
}

/// Loops are exact ports of anyclip.py:1679-1862 / formacOS Watchdogs.swift.
public static class Watchdogs
{
    /// App-layer heartbeat while linked. Two jobs:
    ///  1. Ping every interval, so an actively broken socket surfaces as a
    ///     send failure + EOF.
    ///  2. Enforce a liveness deadline. A half-open socket — the peer slept or
    ///     vanished without RST/FIN — accepts our pings silently and never
    ///     delivers EOF, so the parked receive idles forever and the link is a
    ///     permanent zombie. Detection can't rely on send failures; we require
    ///     inbound traffic (the peer pongs our pings). If nothing arrives for
    ///     interval * deadFactor, the link is dead — drop it so the reconnect
    ///     loop runs. (Field bug: a Mac held a dead link for ~50 min after its
    ///     peer slept, which made the peer idle-bounce forever.)
    public static async Task LinkPingLoopAsync(
        PeerLink link, double intervalSeconds, CancellationToken ct, double deadFactor = 3)
    {
        while (true)
        {
            await Task.Delay(TimeSpan.FromSeconds(intervalSeconds), ct);
            if (!link.IsActive) continue;
            await link.SendPingAsync();
            var idle = link.SecondsSinceInbound();
            if (idle is double s && s > intervalSeconds * deadFactor)
                link.DropStaleLink(s);
        }
    }

    public static async Task NetworkWatchdogAsync(
        IMdnsService mdns, Func<string?> primaryIPv4,
        double intervalSeconds, CancellationToken ct)
    {
        while (true)
        {
            await Task.Delay(TimeSpan.FromSeconds(intervalSeconds), ct);
            var previous = mdns.AdvertisedIp;
            if (previous is null) continue;
            var current = primaryIPv4();
            if (current is not null && current != previous)
                throw new DaemonRestartException(
                    $"local IPv4 changed: {previous} -> {current}; "
                    + "restarting daemon to re-advertise mDNS");
        }
    }

    public static async Task IdleLinkWatchdogAsync(
        IMdnsService mdns, LinkManager manager,
        double idleThresholdSeconds, int refreshAttempts, CancellationToken ct)
    {
        int consecutiveIdle = 0;
        while (true)
        {
            await Task.Delay(TimeSpan.FromSeconds(idleThresholdSeconds), ct);
            // Global escalator: only when the WHOLE mesh is down (zero links).
            if (manager.ActiveLinkCount > 0) { consecutiveIdle = 0; continue; }
            consecutiveIdle++;
            if (consecutiveIdle <= refreshAttempts)
            {
                RotatingLog.Shared.Info(
                    $"no active links for {(int)(idleThresholdSeconds * consecutiveIdle)}s; "
                    + $"refreshing mDNS (attempt {consecutiveIdle}/{refreshAttempts})");
                mdns.Refresh();
            }
            else
            {
                throw new DaemonRestartException(
                    $"no active links with no recovery after {refreshAttempts} mDNS "
                    + "refresh attempts; bouncing daemon");
            }
        }
    }

    public static async Task MdnsReconnectLoopAsync(
        PeerDirectory directory, LinkManager manager, CancellationToken ct)
    {
        double backoff = 1;
        while (true)
        {
            var peers = directory.PeersSnapshot();
            bool attempted = false;
            foreach (var (host, port, label) in peers)
            {
                if (ct.IsCancellationRequested) return;
                if (manager.AtCap) break;                    // no slots: stop dialing this cycle
                if (manager.HasLinkToHost(host)) continue;   // already meshed with this peer
                attempted = true;
                await manager.TryConnectAsync(host, port, label, ct);
                if (manager.HasLinkToHost(host))
                {
                    directory.ClearFails(label);
                    continue;
                }
                int fails = directory.RecordFail(label);
                if (fails >= Wire.MaxReconnectFails)
                {
                    directory.PruneAddress(label);
                    RotatingLog.Shared.Info(
                        $"pruned stale peer address {label} after {fails} failed "
                        + "attempts; awaiting fresh mDNS discovery");
                }
            }
            if (attempted)
            {
                await Task.Delay(TimeSpan.FromSeconds(Math.Min(backoff, 60)), ct);
                backoff = Math.Min(backoff * 2, 60);
            }
            else { backoff = 1; await Task.Delay(TimeSpan.FromSeconds(2), ct); }
        }
    }
}

File: formacOS/Sources/AnyClipDaemon/Watchdogs.swift
import Foundation
import AnyClipCore

/// Thrown by watchdogs to unwind the task group; the in-process supervisor
/// restarts the daemon with backoff (Python: RuntimeError -> supervisor).
public struct DaemonRestartError: Error, CustomStringConvertible {
    public let message: String
    public var description: String { message }
    public init(_ message: String) { self.message = message }
}

func sleepSeconds(_ seconds: Double) async throws {
    try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
}

/// App-layer heartbeat while linked. Two jobs:
///  1. Ping every `interval`s, so an *actively broken* socket surfaces as a
///     send failure + EOF.
///  2. Enforce a liveness deadline. A half-open socket — the peer slept or
///     vanished without RST/FIN — accepts our pings silently and never
///     delivers EOF, so the parked receive idles forever and the link is a
///     permanent zombie. Detection therefore can't rely on send failures; we
///     require *inbound* traffic (the peer pongs our pings). If nothing
///     arrives for `interval * deadFactor`, the link is dead — drop it so the
///     reconnect loop runs. (Field bug: a Mac held a dead link for ~50 min
///     after its peer slept, which in turn made the peer idle-bounce forever.)
public func linkPingLoop(
    link: PeerLink, interval: Double = 30, deadFactor: Double = 3
) async throws {
    while true {
        try await sleepSeconds(interval)
        guard await link.isActive else { continue }
        await link.sendPing()
        if let idle = await link.secondsSinceInbound(), idle > interval * deadFactor {
            await link.dropStaleLink(idleSeconds: idle)
        }
    }
}

/// Bounce the daemon when the host IPv4 changes — the Bonjour advertisement
/// carries the old address and quietly stops working otherwise.
public func networkWatchdog(beacon: MdnsBeacon, interval: Double = 15) async throws {
    while true {
        try await sleepSeconds(interval)
        guard let previous = await beacon.advertisedIP else { continue }
        if let current = primaryIPv4(), current != previous {
            throw DaemonRestartError(
                "local IPv4 changed: \(previous) -> \(current); "
                + "restarting daemon to re-advertise mDNS")
        }
    }
}

/// Self-heal mDNS when NO link is active for too long: refresh browse +
/// re-announce up to `refreshAttempts` times, then bounce the daemon. Keys on
/// the manager's active-link count, never on a single link's idleness — keyed
/// per-link, one sleeping peer would bounce the daemon and tear down every
/// healthy link (spec: global escalator fires only at zero active links).
public func idleLinkWatchdog(
    beacon: MdnsBeacon, manager: LinkManager,
    idleThreshold: Double = 60, refreshAttempts: Int = 3
) async throws {
    var consecutiveIdle = 0
    while true {
        try await sleepSeconds(idleThreshold)
        if await manager.activeLinkCount() > 0 {
            consecutiveIdle = 0
            continue
        }
        consecutiveIdle += 1
        if consecutiveIdle <= refreshAttempts {
            AnyLog.shared.info(
                "no active links for \(Int(idleThreshold * Double(consecutiveIdle)))s; "
                + "refreshing mDNS (attempt \(consecutiveIdle)/\(refreshAttempts))")
            await beacon.refresh()
            await manager.reAnnounce()
        } else {
            throw DaemonRestartError(
                "no active links after \(refreshAttempts) mDNS refresh attempts; "
                + "bouncing daemon")
        }
    }
}

/// Ensure a link to every discovered peer (up to the cap). Keyed per address on
/// the beacon; the manager's node_id table de-dupes so an already-meshed peer is
/// never re-dialed. Re-admission after a drop happens here on the next pass (no
/// new timer). Backoff is per-address via recordFail/pruneAddress.
public func mdnsReconnectLoop(beacon: MdnsBeacon, manager: LinkManager) async throws {
    while true {
        let peers = await beacon.peersSnapshot()
        for peer in peers {
            if await manager.atCap { break }               // no free slots this pass
            if await manager.hasLink(nodeID: peer.peerID) { continue }  // already meshed
            let outcome = await manager.tryConnect(to: peer.endpoint, label: peer.label)
            switch outcome {
            case .routed:
                await beacon.clearFails(label: peer.label)
            case .atCap, .busy:
                break                                       // don't penalise the address
            case .failed:
                let fails = await beacon.recordFail(label: peer.label)
                if fails >= Wire.maxReconnectFails {
                    await beacon.pruneAddress(label: peer.label)
                    AnyLog.shared.info(
                        "pruned stale peer address \(peer.label) after \(fails) failed "
                        + "attempts; awaiting fresh mDNS discovery")
                }
            }
        }
        try await sleepSeconds(2)
    }
}
